src/include/storage/delta_record.h

```diff
@@ -0,0 +1,273 @@
+    PELOTON_ASSERT(offset < num_cols_, "Column offset out of bounds.");
+    Bitmap().Set(offset, false);
+  }
+
+  /**
+   * Set the attribute in the ProjectedRow to be not null using the internal bitmap
+   * @param offset The 0-indexed element to access in this ProjectedRow
+   */
+  void SetNotNull(const uint16_t offset) {
+    PELOTON_ASSERT(offset < num_cols_, "Column offset out of bounds.");
+    Bitmap().Set(offset, true);
+  }
+
+ private:
+  uint32_t size_;
+  uint16_t num_cols_;
+  byte varlen_contents_[0];
+
+  uint32_t *AttrValueOffsets() { return reinterpret_cast<uint32_t *>(ColumnIds() + num_cols_); }
+
+  const uint32_t *AttrValueOffsets() const { return reinterpret_cast<const uint32_t *>(ColumnIds() + num_cols_); }
+
+  common::RawBitmap &Bitmap() { return *reinterpret_cast<common::RawBitmap *>(AttrValueOffsets() + num_cols_); }
+
+  const common::RawBitmap &Bitmap() const {
+    return *reinterpret_cast<const common::RawBitmap *>(AttrValueOffsets() + num_cols_);
+  }
+};
+
+class DataTable;
+/**
+ * Extension of a ProjectedRow that adds two additional fields: a timestamp and a pointer to the next entry in the
+ * version chain
+ */
+class UndoRecord {
+ public:
+  UndoRecord() = delete;
+  DISALLOW_COPY_AND_MOVE(UndoRecord)
+  ~UndoRecord() = delete;
+
+  /**
+   * @return Pointer to the next element in the version chain
+   */
+  std::atomic<UndoRecord *> &Next() { return next_; }
+
+  /**
+   * @return Timestamp up to which the old projected row was visible.
+   */
+  std::atomic<timestamp_t> &Timestamp() { return timestamp_; }
+
+  /**
+   * @return Timestamp up to which the old projected row was visible.
+   */
+  const std::atomic<timestamp_t> &Timestamp() const { return timestamp_; }
+
+  /**
+   * @return the DataTable this DeltaRecord points to
+   */
+  DataTable *Table() const { return table_; }
+
+  /**
+   * @return the TupleSlot this DeltaRecord points to
+   */
+  TupleSlot Slot() const { return slot_; }
+
+  /**
+   * Access the ProjectedRow containing this record's modifications
+   * @return pointer to the delta (modifications)
+   */
+  ProjectedRow *Delta() { return reinterpret_cast<ProjectedRow *>(varlen_contents_); }
+
+  /**
+   * Access the next version in the delta chain
+   * @return pointer to the next version
+   */
+  const ProjectedRow *Delta() const { return reinterpret_cast<const ProjectedRow *>(varlen_contents_); }
+
+  /**
+   * @return size of this DeltaRecord in memory, in bytes.
+   */
+  uint32_t Size() { return static_cast<uint32_t>(sizeof(UndoRecord) + Delta()->Size()); }
+
+  /**
+   * @param redo the redo changes to be applied
+   * @return size of the DeltaRecord which can store the delta resulting from applying redo in memory, in bytes
+   */
+  static uint32_t Size(const ProjectedRow &redo) { return static_cast<uint32_t>(sizeof(UndoRecord)) + redo.Size(); }
+
+  /**
+   * Calculates the size of this DeltaRecord, including all members, values, and bitmap
```

Comment:
Reword to remove DeltaRecord references since we're now talking UndoRecords. Also do we want to rename the file, or are you anticipating the RedoRecord living here too?